### PR TITLE
KAFKA-14937: Refactoring for client code to reduce boilerplate

### PR DESCRIPTION
**Pull request by kirktrue**
There are a number of places in the client code where the same basic calls are made by more than one client implementation. Minor refactoring will reduce the amount of boilerplate code necessary for the client to construct its internal state.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/ClientUtils.java [hunk 1]
```diff
@@ -134,4 +147,112 @@ static List<InetAddress> filterPreferredAddresses(InetAddress[] allAddresses) {
...
+                    sensor,
+                    logContext);
+        } catch (Throwable t) {
+            closeQuietly(selector, "Selector");
+            closeQuietly(channelBuilder, "ChannelBuilder");
+            throw new KafkaException("Failed to create new NetworkClient", t);
+        }
+    }
+
+    public static NetworkClient createNetworkClient(AbstractConfig config,
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/clients/ClientUtils.java:201]: Could this call the other `createNetworkClient()` method below?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        OffsetResetStrategy strategy = OffsetResetStrategy.valueOf(s);
+        return new SubscriptionState(logContext, strategy);
+    }
+
+    public static Metrics metrics(ConsumerConfig config, Time time) {
+        String clientId = config.getString(ConsumerConfig.CLIENT_ID_CONFIG);
+        return metrics(config, time, clientId);
+    }
+
+    public static Metrics metrics(ConsumerConfig config, Time time, String clientId) {
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java]: This is only called from the override above, maybe we can merge both?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchConfig.java [hunk 3]
```diff
@@ -66,8 +67,7 @@
     final int maxPollRecords;
     final boolean checkCrcs;
     final String clientRackId;
-    final Deserializer<K> keyDeserializer;
-    final Deserializer<V> valueDeserializer;
+    final Deserializers<K, V> deserializers;
     final IsolationLevel isolationLevel;
 
     public FetchConfig(int minBytes,
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchConfig.java:73]: It looks like this is only called from tests?

**Review by mimaison — COMMENTED** [review]
Thanks for the PR! The changes look good overall, I left a few minor comments.

The PR includes `KafkaAdminClient.java.orig` and seems to fail spotbugs:
https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13640/1/consoleFull

File: clients/src/main/java/org/apache/kafka/clients/ClientUtils.java (hunk 1 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/ClientUtils.java:201]: Not directly, no. I created another version of `createNetworkClient` that the other two can call, though.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchConfig.java (hunk 3 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchConfig.java:73]: No, it's really only called by `CompletedFetch`. It's just that we have to pass the two deserializers around through multiple layers of code.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java (hunk 2 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java]: Done.

File: clients/src/main/java/org/apache/kafka/clients/ClientUtils.java (hunk 1 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/ClientUtils.java:201]: Done.

**Comment by kirktrue** [comment]
Thanks for the review @mimaison!

I removed the `KafkaAdminClient.java.orig` file. I believe I have resolved the Spotbugs failure, but I will wait for the Jenkins build to make sure it's not just fixed on my machine 😄

**Comment by kirktrue** [comment]
The Spotbugs validation is passing and there are no new test failures.

**Review by mimaison — APPROVED** [review]
LGTM, thanks

**Comment by kirktrue** [comment]
@mimaison Would you be willing to merge this change? Thanks!

**Comment by mimaison** [comment]
@kirktrue I'm looking at it now. With this PR I consistently get failures with the following tests:
```
SslAdminIntegrationTest.testAclUpdatesUsingAsynchronousAuthorizer()
SslAdminIntegrationTest.testAclUpdatesUsingSynchronousAuthorizer()
```
Both have the same failure:
```
org.opentest4j.AssertionFailedError: Invalid client id:  ==> expected: <true> but was: <false>
	at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at app//org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
	at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
	at app//org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:211)
	at app//kafka.api.SslAdminIntegrationTest.validateRequestContext$1(SslAdminIntegrationTest.scala:203)
```
These tests work fine on trunk. Any idea what could be the cause?

**Comment by kirktrue** [comment]
> With this PR I consistently get failures with the following tests

Thanks for catching that. I was not setting the client ID to the correct value for the `KafkaAdminClient`'s `NetworkClient`.

**Comment by kirktrue** [comment]
Closing pull request and moving to a fork where we're doing integration.

**Pull request closed without merging**